panic: deadlock detected at fmut

Possibly related to https://github.com/syncthing/syncthing/issues/7230, but not sure. It happened 15 minutes before the other panics. The cause is unknown, but I was maybe trying to pause and resume a folder at that time.

```
[D4DZU] 18:55:26 INFO: Failed to exchange Hello messages with CWXZ7CV-XXXXXXX-XXXXXXX-XXXXXXX-XXXXXXX-XXXXXXX-XXXXXXX-XXXXXXX at [::]:22000-178.42.98.249:22000/quic-server/TLS1.3-TLS_AES_128_GCM_SHA256: Application error 0x0: closing
Panic at 2020-12-24T19:08:28+09:00
panic: deadlock detected at fmut

goroutine 58 [running]:
github.com/syncthing/syncthing/lib/model.(*deadlockDetector).Watch.func1(0x12c60ff0, 0x32f86fd8, 0x12baa720, 0xd961dc, 0x4)
	github.com/syncthing/syncthing/lib/model/util.go:61 +0x2b4
created by github.com/syncthing/syncthing/lib/model.(*deadlockDetector).Watch
	github.com/syncthing/syncthing/lib/model/util.go:37 +0x93

goroutine 1 [chan receive, 55 minutes]:
github.com/syncthing/syncthing/lib/syncthing.(*App).Wait(...)
	github.com/syncthing/syncthing/lib/syncthing/syncthing.go:370
main.syncthingMain(0x12a58700, 0x1f, 0x0, 0x0, 0x0, 0x1, 0x0, 0x0, 0x0, 0x0, ...)
	github.com/syncthing/syncthing/cmd/syncthing/main.go:719 +0x59f
main.main()
	github.com/syncthing/syncthing/cmd/syncthing/main.go:438 +0x33b
```

[panic-20201224-190828.log](https://github.com/syncthing/syncthing/files/5739282/panic-20201224-190828.log)

Syncthing v1.12.0.

**Edit**: Just for the record, I have found the older issue (https://github.com/syncthing/syncthing/issues/6076), but this here is not a slow device.

--- a/lib/model/indexsender.go
+++ b/lib/model/indexsender.go
@@ -30,6 +30,7 @@ type indexSender struct {
 	prevSequence             int64
 	evLogger                 events.Logger
 	connClosed               chan struct{}
+	done                     chan struct{}
 	token                    suture.ServiceToken
 	pauseChan                chan struct{}
 	resumeChan               chan *db.FileSet
@@ -38,6 +39,7 @@ type indexSender struct {
 func (s *indexSender) Serve(ctx context.Context) (err error) {
 	l.Debugf("Starting indexSender for %s to %s at %s (slv=%d)", s.folder, s.conn.ID(), s.conn, s.prevSequence)
 	defer func() {
+		close(s.done)
 		err = svcutil.NoRestartErr(err)
 		l.Debugf("Exiting indexSender for %s to %s at %s: %v", s.folder, s.conn.ID(), s.conn, err)
 	}()
@@ -101,14 +103,14 @@ func (s *indexSender) Serve(ctx context.Context) (err error) {
 
 func (s *indexSender) resume(fset *db.FileSet) {
 	select {
-	case <-s.connClosed:
+	case <-s.done:
 	case s.resumeChan <- fset:
 	}
 }
 
 func (s *indexSender) pause() {
 	select {
-	case <-s.connClosed:
+	case <-s.done:
 	case s.pauseChan <- struct{}{}:
 	}
 }
@@ -314,6 +316,7 @@ func (r *indexSenderRegistry) addLocked(folder config.FolderConfiguration, fset 
 	is := &indexSender{
 		conn:                     r.conn,
 		connClosed:               r.closed,
+		done:                     make(chan struct{}),
 		folder:                   folder.ID,
 		folderIsReceiveEncrypted: folder.Type == config.FolderTypeReceiveEncrypted,
 		fset:                     fset,
@@ -366,7 +369,7 @@ func (r *indexSenderRegistry) removeAllExcept(except map[string]struct{}) {
 			delete(r.indexSenders, folder)
 		}
 	}
-	for folder := range r.indexSenders {
+	for folder := range r.startInfos {
 		if _, ok := except[folder]; !ok {
 			delete(r.startInfos, folder)
 		}

